mvebu: improve thermal management of IEI Puzzle devices

dangowrt opened the pull request:
- Make step_wise thermal governor respect hysteresis This is done by importing a downstream patch, backporting the same feature now present in Linux v6.10+ would be too messy.
 - Introduce thermal zone for the WT61P803 uC (chassis and board sensors)
 - Introduce thermal zones for AQR NBase-T PHYs
 - No longer modify existing SoC thermal zones (which are now only in charge for emergency shutdown, and can be interrupt driven instead of polled)

Due to a bug in Linux 6.1 there are error messages because the thermal subsystems tries to program trip-points near absolute 0 temperature which the hardware doesn't support. Once the system has been running for a while and the first trip points have been passed those messages no longer occur. They don't seem to impact functionality, so I'm a bit reluctant to put more work into Linux 6.1 which anyway is never going to be part of a release.

dangowrt committed:
mvebu: improve thermal management of IEI Puzzle devices

 - Make step_wise thermal governor respect hysteresis
   This is done by importing a downstream patch, backporting the same feature
   now present in Linux v6.10+ would be too messy.
 - Introduce thermal zone for the WT61P803 uC (chassis and board sensors)
 - Introduce thermal zones for AQR NBase-T PHYs
 - No longer modify existing SoC thermal zones (which are now only in charge
   for emergency shutdown, and can be interrupt driven instead of polled)

Signed-off-by: Daniel Golle <daniel@makrotopia.org>
Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

zcracker commented:
[  425.115896] thermal thermal_zone3: Failed to set trips: -34
[  427.207644] thermal thermal_zone3: Failed to set trips: -34
[  429.301451] thermal thermal_zone3: Failed to set trips: -34
[  437.658669] thermal thermal_zone3: Failed to set trips: -34
[  443.930622] thermal thermal_zone3: Failed to set trips: -34
[  447.074821] thermal thermal_zone3: Failed to set trips: -34
[  451.254556] thermal thermal_zone3: Failed to set trips: -34
[  455.432186] thermal thermal_zone3: Failed to set trips: -34
[  459.598526] thermal thermal_zone3: Failed to set trips: -34
[  464.816976] thermal thermal_zone3: Failed to set trips: -34
[  466.913730] thermal thermal_zone3: Failed to set trips: -34
[  470.043661] thermal thermal_zone3: Failed to set trips: -34
[  472.130184] thermal thermal_zone3: Failed to set trips: -34
[  479.420650] thermal thermal_zone3: Failed to set trips: -34
[  488.814148] thermal thermal_zone3: Failed to set trips: -34

dangowrt commented:
@zcracker Yes, see patch description. Not really a problem and happens only on Linux 6.1. Use Linux 6.6 and the errors are gone.

zcracker commented:
SYNC    include/config/auto.conf.cmd
*
* Restart config...
*
*
* Performance monitor support
*
ARM CCI PMU driver (ARM_CCI_PMU) [N/m/y/?] n
ARM CCN driver support (ARM_CCN) [N/m/y/?] n
Arm CMN-600 PMU support (ARM_CMN) [N/m/y/?] n
ARM PMU framework (ARM_PMU) [Y/n/?] y
ARM SMMUv3 Performance Monitors Extension (ARM_SMMU_V3_PMU) [N/m/y/?] (NEW) 
ARM DynamIQ Shared Unit (DSU) PMU (ARM_DSU_PMU) [N/m/y/?] n
Enable support for the ARMv8.2 Statistical Profiling Extension (ARM_SPE_PMU) [N/m/y/?] n
HiSilicon PCIE PERF PMU (HISI_PCIE_PMU) [N/m/y/?] n
HNS3 PERF PMU (HNS3_PMU) [N/m/y/?] n
ARM Coresight Architecture PMU (ARM_CORESIGHT_PMU_ARCH_SYSTEM_PMU) [N/m/y/?] (NEW)

zcracker commented:
kernel 6.6.29 
[ 1840.242623] thermal thermal_zone11: failed to read out thermal zone (-5)

zcracker commented:
> SYNC include/config/auto.conf.cmd *
> 
> * Restart config...
> * Performance monitor support
> 
> ARM CCI PMU driver (ARM_CCI_PMU) [N/m/y/?] n ARM CCN driver support (ARM_CCN) [N/m/y/?] n Arm CMN-600 PMU support (ARM_CMN) [N/m/y/?] n ARM PMU framework (ARM_PMU) [Y/n/?] y ARM SMMUv3 Performance Monitors Extension (ARM_SMMU_V3_PMU) [N/m/y/?] (NEW) ARM DynamIQ Shared Unit (DSU) PMU (ARM_DSU_PMU) [N/m/y/?] n Enable support for the ARMv8.2 Statistical Profiling Extension (ARM_SPE_PMU) [N/m/y/?] n HiSilicon PCIE PERF PMU (HISI_PCIE_PMU) [N/m/y/?] n HNS3 PERF PMU (HNS3_PMU) [N/m/y/?] n ARM Coresight Architecture PMU (ARM_CORESIGHT_PMU_ARCH_SYSTEM_PMU) [N/m/y/?] (NEW)

/openwrt/target/linux/generic/config-6.6 
add  
# CONFIG_ARM_SMMU_V3_PMU is not set
# CONFIG_ARM_CORESIGHT_PMU_ARCH_SYSTEM_PMU is not set
Compilation passed

zcracker commented:
> kernel 6.6.29 [ 1840.242623] thermal thermal_zone11: failed to read out thermal zone (-5)

It seems that the fan is always at 3420 rpm

zcracker commented:
Test ok, please merge tks

dangowrt commented:
> kernel 6.6.29 [ 1840.242623] thermal thermal_zone11: failed to read out thermal zone (-5)

That's strange though. Is that the puzzle_mcu thermal sensor? Or one of the PHYs? Can be that temperature of the PHYs can only be read when the interface is up, we may want to add some logic around that then.

zcracker commented:
[22725.678066] iei-wt61p803-puzzle serial0-0: Command reply receive timeout
[22725.684829] thermal thermal_zone12: failed to read out thermal zone (-110)
[22930.114812] thermal thermal_zone12: failed to read out thermal zone (-5)
[24481.097307] iei-wt61p803-puzzle serial0-0: Command reply receive timeout
[24481.104064] thermal thermal_zone12: failed to read out thermal zone (-110)
[29474.193014] thermal thermal_zone12: failed to read out thermal zone (-5)
[33913.901629] thermal thermal_zone12: failed to read out thermal zone (-5)
[34282.477319] thermal thermal_zone11: failed to read out thermal zone (-5)

zcracker commented:
> > kernel 6.6.29 [ 1840.242623] thermal thermal_zone11: failed to read out thermal zone (-5)
> 
> That's strange though. Is that the puzzle_mcu thermal sensor? Or one of the PHYs? Can be that temperature of the PHYs can only be read when the interface is up, we may want to add some logic around that then.


This error occurs after running for a period of time

Ansuel committed:
mvebu: fix missing property in puzzle thermal

Fix missing property in puzzle thermal. The thing was never supposed to
work.

Property #thermal-sensor-cells was missing from the puzzle hwmon, making
the entire thermal platform referencing that fail to probe with -EINVAL.

The puzzle hwmon expose 2 termistor but they probably use an userspace
downstream utility to configure and handle thermal. For this reason we
really don't know what they use the sensor for or when it's attached.

We use them to sensor if the Chassis gets too hot due to ambient
temperature and generic components getting too warm.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel committed:
mvebu: split thermal zone for puzzle chassis

Split thermal zone for puzzle chassis. Thermal platform supports only
one sensor per thermal zone.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>
Signed-off-by: Daniel Golle <daniel@makrotopia.org>

Ansuel committed:
mvebu: disable polling delay for passive trip point for puzzle thermal

We don't have any passive trip point hence we can set the polling delay
for passive trip to 0 effectively disabling this polling.

Signed-off-by: Christian Marangi <ansuelsmth@gmail.com>

Ansuel commented:
@dangowrt sorry for the email spam for the 10000 hard push... please don't hate me...

dangowrt commented:
Thank you @Ansuel for taking care of it!
Another two force-push on top from my side which will hopefully make it build now (there was still a line-wrapping issue in puzzle-thermal.dtsi)
